appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartData.java

```diff
@@ -93,6 +104,134 @@ private void setElementsFromPairsProperty(String text){
         chartDataModel.setElements(currentElements);
     }
 
+    /**
+     * Sets the Source property of the Chart Data component.
+     * Responsible for showing/hiding properties according to the
+     * attached Source component type.
+     *
+     * @param source  Name of the new Source component attached
+     */
+    private void setSourceProperty(String source) {
+        // If a Data Source was previously assigned and is of type CSVFile,
+        // de-attach this Data component from it
+        if (dataSource instanceof MockCSVFile) {
+            ((MockCSVFile)dataSource).removeCSVFileChangeListener(this);
+        }
+
+        // Get the newly attached Source component
+        dataSource = editor.getComponents().getOrDefault(source, null);
+
+        // Hide Elements from Pairs property if a Data Source has been set
+        showProperty(PROPERTY_PAIRS, (dataSource == null));
+
+        // Handle CSV-related property responses
+        handleCSVPropertySetting();
+
+        // If the Data Source is now null, set back the
+        // currentElements property. An additional check
+        // is required to check whether the Data component is
+        // attached (prevents some exceptions with regards to
+        // removing the CSVFile component)
+        if (dataSource == null && isAttached()) {
```

Comment:
Related bug (to isAttached()): Attach Source to Data component, remove it, change Source file of the CSVFile